lib/SILGen/SILGenStmt.cpp

```diff
@@ -932,10 +934,276 @@ void StmtEmitter::visitRepeatWhileStmt(RepeatWhileStmt *S) {
   SGF.BreakContinueDestStack.pop_back();
 }
 
+namespace {
+  class CancelCleanup : public Cleanup {
+    SILLocation loc;
+    std::function<ArgumentSource ()> iteratorGen;
+    ConcreteDeclRef generatorCancelRef;
+  public:
+    CancelCleanup(SILLocation loc, std::function<ArgumentSource ()> iteratorGen, 
+      ConcreteDeclRef generatorCancelRef) : 
+      loc(loc), iteratorGen(iteratorGen),
+      generatorCancelRef(generatorCancelRef) { }
+
+    void emit(SILGenFunction &SGF, CleanupLocation l, ForUnwind_t forUnwind) override {
+      FormalEvaluationScope scope(SGF);
+      SGF.emitApplyMethod(loc, generatorCancelRef, iteratorGen(),
+          PreparedArguments(ArrayRef<AnyFunctionType::Param>({})),
+          SGFContext());
+    }
+
+    void dump(SILGenFunction &) const override {
+#ifndef NDEBUG
+      llvm::errs() << "CancelCleanup\n"
+                   << "State: " << getState() << "\n";
+#endif
+    }
+  };
+}
+
+void StmtEmitter::visitAsyncForEachStmt(ForEachStmt *S) {
+
+  // Dig out information about the sequence conformance.
+  auto sequenceConformance = S->getSequenceConformance();
+  Type sequenceType = S->getSequence()->getType();
+  
+  auto asyncSequenceProto =
+      SGF.getASTContext().getProtocol(KnownProtocolKind::AsyncSequence);
+  auto sequenceSubs = SubstitutionMap::getProtocolSubstitutions(
+      asyncSequenceProto, sequenceType, sequenceConformance);
+
+  // Emit the 'generator' variable that we'll be using for iteration.
+  LexicalScope OuterForScope(SGF, CleanupLocation(S));
+  {
+    auto initialization =
+        SGF.emitInitializationForVarDecl(S->getIteratorVar(), false);
+    SILLocation loc = SILLocation(S->getSequence());
+
+    // Compute the reference to the AsyncSequence's makeAsyncSequence().
+    FuncDecl *makeGeneratorReq = 
+      SGF.getASTContext().getAsyncSequenceMakeAsyncIterator();
```

Comment:
Can you pull SGF.getASTContext() into a local variable? It is used in several places here so it would make the code a bit easier to follow